lib/parser/message.h

```diff
@@ -240,13 +240,24 @@ class Messages {
 
 class ContextualMessages {
 public:
+  class SavedState {
```

Comment:
If I understand, you expect this code to be used like

```
{ auto namedoesntmatter{context.PushLocation(here)};
  ...
}
```

I was thinking along the lines of something like:

```
{ WithMessageLocation unusedname{context, here};
  ....
}
```

Modulo naming, I'm not sure which is more clear.  (Maybe "Push" shouldn't be in the name since there's no explicit "Pop".)